examples/spouts/go-rabbitmq-spout/source/main.go

```diff
@@ -0,0 +1,408 @@
+package main
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/streadway/amqp"
+
+	client "github.com/pachyderm/pachyderm/src/client"
+)
+
+const defaultPrefetch = 500
+const defaultFlushInterval = 10000 // ms
+const defaultExtension = "ndjson"
+const defaultBranch = "staging"
+const defaultSwitchBranch = "master"
+const defaultSwitchInterval = 60000 // ms
+
+var logger log.Logger
+
+func writeFiles(pc *client.APIClient, opts *options, buffer []amqp.Delivery, commits chan<- string) error {
+
+	log.Print("writing buffer...")
+	log.Printf("buffer_size %d", len(buffer))
+
+	// hash messages to ensure there are no duplicated
+	var maxMessage amqp.Delivery
+	msgMap := make(map[string]amqp.Delivery)
+
+	// groupby hash
+	groupHash := md5.New()
+
+	// Hash messages to ensure there are no duplicates
+	for _, msg := range buffer {
+		hash := md5.New()
+		hash.Write(msg.Body)
+		groupHash.Write(msg.Body)
+		// Hash to ensure uniqueness of the messages
+		name := hex.EncodeToString(hash.Sum(nil))
+
+		// Keep track of the highest delivery tag
+		if msg.DeliveryTag > maxMessage.DeliveryTag {
+			maxMessage = msg
+		}
+
+		msgMap[name] = msg
+	}
+
+	// Give a warning if we found duplicates. Not a big deal
+	if len(msgMap) != len(buffer) {
+		log.Printf("duplicate messages in buffer: %d vs %d", len(msgMap), len(buffer))
+	}
+
+	// Start of writing
+	log.Print("Writing messages...")
+
+	// Begin the commit
+	commit, err := pc.StartCommit(opts.repoName, opts.commitBranch)
+	if err != nil {
+		// StartCommit will fail if this crashed with an open commit. Finish open commits and
+		commitList, err := pc.ListCommit(opts.repoName, opts.commitBranch, "", 0)
+		for _, c := range commitList {
+			com := c.GetCommit()
+			err = pc.FinishCommit(opts.repoName, com.GetID())
+			if err != nil {
+				return fmt.Errorf("unable to start commit on %s@%s: %v", opts.repoName, opts.commitBranch, err)
+			}
+		}
+		if err != nil {
+			return fmt.Errorf("unable to start commit on %s@%s: %v", opts.repoName, opts.commitBranch, err)
+		}
+	}
```

Comment:
This cleanup code can be removed if you do an atomic commit instead of StartCommit and FinishCommit. all you need to do is call `pc.NewPutFileClient()` and you'll get back a client that you can do `PutFile*`s operations on like normal and call `Close()` on it when you're done. If you crash midway through the operation everything will be rolled back.

Reply:
Awesome. I'll make that change and test it out.

Reply:
Hey Daniel,

I set up a AWS MQ - RabbitMQ to connect from Pach Hub. It worked like a charm but only supports amqps. Added it to the code. It felt a little too much to fork your repo and push a PR in your PR branch. ;-) Feel free to patch/adapt directly or I can do it after the merge. LGTM

```

diff --git a/examples/spouts/go-rabbitmq-spout/pipelines/spout.pipeline.json b/examples/spouts/go-rabbitmq-spout/pipelines/spout.pipeline.json
index 7549aee50..12e5ca5ae 100644
--- a/examples/spouts/go-rabbitmq-spout/pipelines/spout.pipeline.json
+++ b/examples/spouts/go-rabbitmq-spout/pipelines/spout.pipeline.json
@@ -17,6 +17,7 @@
             "SWITCH_INTERVAL_MS": "60000",
             "EXTENSION": "ndjson",
             "TEST": "false",
+            "RABBITMQ_PROTOCOL": "amqps",
             "RABBITMQ_HOST": "rabbitmq.default.svc.cluster.local:5672",
             "RABBITMQ_USER": "peter",
             "SWITCH_BRANCH": "master",
diff --git a/examples/spouts/go-rabbitmq-spout/source/main.go b/examples/spouts/go-rabbitmq-spout/source/main.go
index d63b472e2..87a16484f 100644
--- a/examples/spouts/go-rabbitmq-spout/source/main.go
+++ b/examples/spouts/go-rabbitmq-spout/source/main.go
@@ -311,7 +311,12 @@ func dialRabbit() (*amqp.Connection, error) {
                return nil, fmt.Errorf("RABBITMQ_PASSWORD environment variable not set")
        }

-       return amqp.Dial("amqp://" + rabbitUser + ":" + rabbitPassword + "@" + rabbitURL)
+       rabbitProtocol := os.Getenv("RABBITMQ_PROTOCOL")
+       if rabbitProtocol == "" {
+               rabbitProtocol = "amqp"
+       }
+
+       return amqp.Dial(rabbitProtocol + "://" + rabbitUser + ":" + rabbitPassword + "@" + rabbitURL)
 }

 func main() { 

```